File: frontend/src/types/availability.ts
import type { DecimalString } from '@/types/analytics'

/**
 * The availability contract, transcribed from `app.schemas.availability` and confirmed
 * against the live OpenAPI document and real searches.
 *
 * Six facts that check established, each of which shaped the UI:
 *
 * 1. **The stay is half-open, `[check_in, check_out)`.** Verified against a real seeded
 *    booking: room 205 is busy for 17→18 September, free for 16→17, and free again for a
 *    stay *beginning* on the 18th. Nothing here re-derives that; the dates go to the server
 *    as typed.
 * 2. **`sufficient` is the answer, and it is the server's.** It is not "the list is
 *    non-empty": a mixed request for `DLX:2` and `STD:99` came back with **both types
 *    listed, sixteen rooms free, and `sufficient: false`**, because one share could not be
 *    met. A UI reading the list instead of the flag would report the opposite of the truth.
 * 3. **Zero availability is a 200 with an empty list**, never a 404. "Nothing free on those
 *    dates" is a fact about the dates, not a failed request.
 * 4. **The two request forms are alternatives.** `rooms` names types and counts; it cannot
 *    be combined with `room_type_code` or `rooms_required` -- both are 422s, verified. The
 *    interface therefore offers a mode, not a pile of fields that can contradict each other.
 * 5. **`room_type_code` is case-sensitive; `rooms` is not.** `room_type_code=dlx` is a
 *    **404**, while `rooms=dlx:1` succeeds -- the mixed parser upper-cases and the single
 *    filter does not. So a code is always sent exactly as the catalogue gave it.
 * 6. **A result is not a reservation.** The schema says so in as many words: nothing is held,
 *    and a room named here may be taken a moment later. The screen repeats it.
 */

/** One physical room the hotel could allocate for the requested stay. */
export interface AvailableRoom {
  /** Unique per hotel. The same identifier the room endpoints put in their paths. */
  readonly room_number: string
  readonly floor: number | null
}

/** A room type, and the rooms of it that are free for the requested stay. */
export interface AvailableRoomType {
  readonly code: string
  readonly name: string
  readonly max_occupancy: number
  readonly standard_occupancy: number
  /** A `Decimal`, so a string on the wire. Formatted for display, never used in arithmetic. */
  readonly base_price: DecimalString
  readonly currency: string
  /**
   * How many rooms of this type are free.
   *
   * The server derives it from the actual allocatable rooms rather than a stored figure.
   * Nothing in this application recomputes it, and nothing compares it to a requested count
   * to decide an outcome -- `sufficient` already carries that answer.
   */
  readonly available_count: number
  /**
   * How many of this type the caller asked for.
   *
   * Present only when the request named types and counts. `null` for a search that named
   * none, because there was no per-type request to report.
   */
  readonly requested_count: number | null
  readonly rooms: readonly AvailableRoom[]
}

/** The answer to one availability search. */
export interface AvailabilityResult {
  readonly hotel_public_id: string
  readonly check_in: string
  readonly check_out: string
  /** `check_out - check_in`, and never zero: the server refuses a stay of no nights. */
  readonly nights: number
  /** How many rooms of a single type were asked for. 1 by default, and 1 for a mixed search. */
  readonly rooms_required: number
  /**
   * **The verdict.** Whether the request can be met.
   *
   * For a single-type search: whether some type can supply `rooms_required` for the whole
   * stay. For a mixed search: whether *every* named type can supply its own share. `false`
   * is a successful answer, not an error -- and it can arrive alongside a non-empty list of
   * types, which is exactly why this flag is read rather than the list's length.
   */
  readonly sufficient: boolean
  /** Free rooms across the types listed. The server's count. */
  readonly available_rooms: number
  readonly room_types: readonly AvailableRoomType[]
}

/** One entry of a mixed request: a room type and how many of it are wanted. */
export interface RoomTypeDemand {
  readonly code: string
  readonly count: number
}

File: backend/app/schemas/availability.py
"""What a hotel could sell for a requested stay.

Stage 4.5.10. A read-only projection of rooms that no inventory-holding allocation overlaps.

**Identifiers follow the ones this API already uses.** A room is named by ``room_number``,
which is unique per hotel (``uq_rooms_hotel_id_room_number``) and is what the rooms endpoints
put in their paths; a room type by its ``code``. Neither has a UUID, so none is invented here,
and no internal BIGINT is exposed.

**A request for several rooms is a request for several rooms of ONE type** (Stage
4.5.25). ``rooms_required`` narrows the answer to types that can supply the whole
party from their own inventory; mixing types to make up a number is a different
question -- one about allocation preferences rather than availability -- and is not
answered here.

**A result is still not a reservation.** Nothing is held, nothing is written, and a
room named here may be taken by another request a moment later. The booking
transaction and its exclusion constraint remain the only authority on inventory.

**The stay interval is half-open**, ``[check_in, check_out)``, matching booking creation and
the exclusion constraint exactly: a stay from the 10th to the 14th occupies the nights of the
10th, 11th, 12th and 13th, and a stay beginning on the 14th does not collide with it. Note
that this differs from the analytics date range, whose bounds are both inclusive -- the two
answer different questions and each matches the thing it describes.
"""

from __future__ import annotations

import datetime as dt
import decimal
import uuid

from pydantic import BaseModel, ConfigDict, Field


class AvailableRoom(BaseModel):
    """One physical room that could be allocated for the requested stay."""

    model_config = ConfigDict(from_attributes=True)

    room_number: str
    floor: int | None = None


class AvailableRoomType(BaseModel):
    """A room type, and the rooms of it that are free for the requested stay.

    ``available_count`` is ``len(rooms)`` -- derived from the actual allocatable rooms rather
    than from a stored inventory figure, because a count kept anywhere else would be a second
    answer to the same question.
    """

    model_config = ConfigDict(from_attributes=True)

    code: str
    name: str
    max_occupancy: int
    standard_occupancy: int
    base_price: decimal.Decimal
    currency: str
    available_count: int
    #: How many of this type the caller asked for, when the request named types and
    #: counts (Stage 4.5.26). Absent for a search that named none, because there was
    #: no per-type request to report.
    requested_count: int | None = None
    rooms: list[AvailableRoom]


class AvailabilityResult(BaseModel):
    """The answer to one availability search.

    An empty ``room_types`` is a successful answer meaning "nothing free", not an error.
    """

    model_config = ConfigDict(from_attributes=True)

    hotel_public_id: uuid.UUID
    check_in: dt.date
    check_out: dt.date
    #: Half-open, so this is ``(check_out - check_in).days`` and never zero.
    nights: int = Field(description="Nights in the requested stay: check_out - check_in.")
    #: How many rooms of a SINGLE type the caller asked for (Stage 4.5.25).
    #: One by default, which is the question this endpoint answered before.
    rooms_required: int = 1
    #: Whether at least one room type can supply ``rooms_required`` rooms for the
    #: whole stay. False is a successful answer meaning the hotel cannot.
    sufficient: bool
    #: Free rooms across the types listed below -- which, when more than one room
    #: is required, are only the types that can supply the whole request.
    available_rooms: int
    room_types: list[AvailableRoomType]


__all__ = ["AvailabilityResult", "AvailableRoom", "AvailableRoomType"]
